Data/Mod/Word.hs

```diff
@@ -217,33 +221,77 @@ instance KnownNat m => Ring (Mod m) where
   negate = P.negate
   {-# INLINE negate #-}
 
--- | Division by a residue, which is not
--- <https://en.wikipedia.org/wiki/Coprime_integers coprime>
--- with the modulus, throws 'DivideByZero'.
--- Consider using 'invertMod' for non-prime moduli.
+-- | 'Mod' @m@ is not even an
+-- <https://en.wikipedia.org/wiki/Integral_domain integral domain> for
+-- <https://en.wikipedia.org/wiki/Composite_number composite> @m@,
+-- much less a <https://en.wikipedia.org/wiki/GCD_domain GCD domain>.
+-- However, 'Data.Euclidean.gcd' and 'Data.Euclidean.lcm' are still meaningful
+-- even for composite @m@, corresponding to a sum and an intersection of
+-- <https://en.wikipedia.org/wiki/Ideal_(ring_theory) ideals>.
+--
+-- The instance is lawful only for
+-- <https://en.wikipedia.org/wiki/Prime_number prime> @m@, otherwise
+-- @'Data.Euclidean.divide' x y@ tries to return any @Just z@ such that @x == y * z@.
+--
 instance KnownNat m => GcdDomain (Mod m) where
-  divide x y = Just $! x / y
-  gcd        = const $ const 1
-  lcm        = const $ const 1
-  coprime    = const $ const True
+  divide (Mod 0) !_ = Just (Mod 0)
```

Comment:
Shall we adjust strictness of `Data.Mod.divide` as well?

Reply:
I figured it doesn't really matter, since `Data.Mod.Mod` is boxed anyway. My naive benchmark doesn't show any improvement either.